Entropy UDF

atavory opened the pull request:
Following #12700  and talks with @rongrong, I'd like to ask to pull the entropy UDF.

Motivation
------------

[Entropy](https://en.wikipedia.org/wiki/Entropy_(information_theory)) is a measure of disorder and uncertainty. 

Say you have a number of web pages, each accessed by approximately 100 visits last month. The visits of one page might be "organized", in the sense that all 100 accesses came from a small number of users, while the visits of another page might be "disorganized", in the sense that the 100 accesses were distributed over many more users. In this case, entropy quantifies how "disorganized" the accesses are, by taking a column of counts per user (summing up to about 100, in this example), and giving a single number.

Entropy is also used as the basis of several other calculations in machine learning and statistics.

Possible inputs and output
------------------------------

  entropy(BIGINT) -> DOUBLE
  entropy(INT) -> DOUBLE

Description
------------

Compute the normalized entropy of a series of counts.  The input is assumed
 to be a column of counts of each value's occurrence . The entropy is normalized in the sense that
 the counts are first normalized so that the sum of all counts equals one
 (i.e., it is converted to a probability distribution).  Any NULLs in the
 column are ignored.  If the column contains negative values then NULL is
 returned.  If the total count of entries of in the column is zero then zero
 is returned.

 Note that the entropy is computed using base-2 log.

atavory commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/EntropyAggregation.java:
```diff
@@ -0,0 +1,78 @@
+        if (count == 0) {
+            return;
+        }
+        if (count < 0) {
```
I think it is in the description above, no? If you could point out an example of a documentation for things like this, I'd appreciate it (I couldn't find any), and I'll do it from there.

atavory commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/EntropyState.java:
```diff
@@ -0,0 +1,32 @@
+
+    boolean getNull();
+
+    void setSumC(double sumC);
```
Added

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/EntropyAggregation.java:
```diff
@@ -0,0 +1,98 @@
+    @OutputFunction(StandardTypes.DOUBLE)
+    public static void output(@AggregationState EntropyState state, BlockBuilder out)
+    {
+        final Double entropy = EntropyAggregation.calculateEntropy(state);
```
We don't normally use the `final` keyword in situations like this. You can remove it.

rongrong replied:
Not sure why you want to have a separate method for this. But if you think it's better to have a separate method, import static and use `Double entropy = calculateEntropy(state)`. Presto convention is to use static import unless it's too confusing. If no other class will call `calculateEntropy`, it should be a private method.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/EntropyAggregation.java:
```diff
@@ -0,0 +1,98 @@
+ *      where $p_i = {c_i \over \sum_i [ c_i ]}$. If $\sum_i [ c_i ] = 0$, the entropy is defined to be 0.
+ */
+@AggregationFunction("entropy")
+@Description("Computes the log-2 entropy of counts inputs")
```
Let's just make sure to mention it's normalized to sum = 1 in the description as well.

atavory commented:
@rongrong Many thanks for your review!

I've addressed your comments. Please note that I've also made the function throw an exception if inputs are illegal, as other UDFs do this too. Have added tests for this as well.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/EntropyAggregation.java:
```diff
@@ -0,0 +1,87 @@
+        if (count < 0) {
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    EntropyAggregation.NEGATIVE_ARGUMENT_MESSAGE);
```
You can inline the message here. There's no need to have a separate variable.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/EntropyAggregation.java:
```diff
@@ -0,0 +1,87 @@
+
+    private static final String NEGATIVE_ARGUMENT_MESSAGE = "Entropy count argument must be non-negative";
+
+    private EntropyAggregation() {}
```
We normally keep constructors at the beginning of the class.

rongrong commented on presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestEntropyAggregation.java:
```diff
@@ -0,0 +1,177 @@
+        return ImmutableList.of(StandardTypes.INTEGER);
+    }
+
+    private static final String FUNCTION_NAME = "entropy";
```
Move this to the beginning of the class, before `private InternalAggregationFunction entropyFunction;`, and keep a blank line between the two.
